Add translation of fpclass llvm intrinsic

MrSidims opened the pull request:
Optimized LLVM IR started to contain this intrinsic recently after InstCombine updates, like:
"Fold and/or of fcmp into class" ( 08f03887 )

There is no direct counterpart for it in SPIR-V, so testing is mapped on a sequence of SPIR-V instructions:
1. test for both qnan and snan -> OpIsNan
2. test for either of qnan and snan -> isquiet(V) ==> abs(V) >= (unsigned(Inf) | quiet_bit)
                                                             issignaling(V) ==> isnan(V) && !isquiet(V)
3. test for neg/pos inf -> OpIsInf + OpSignBitSet
4. test for neg/pos normal -> OpIsNormal + OpSignBitSet
5. test for neg/pos subnormal -> issubnormal(V) ==> unsigned(abs(V) - 1) < (all mantissa bits set)
6. test for neg/pos zero -> compare bitcasted to int float with 0 + OpSignBitSet

- [x] add test

Commit:
Add translation of fpclass llvm intrinsic

There is no direct counterpart for it in SPIR-V, so testing is mapped
on a sequence of SPIR-V instructions:
1. test for qnan and snan -> OpIsNan
2. test for neg/pos inf -> OpIsInf + OpSignBitSet
3. test for neg/pos normal -> OpIsNormal + OpSignBitSet
4. test for neg/pos subnormal -> not (OpIsNormal or OpIsNan) +
   OpSignBitSet
5. test for neg/pos zero -> OpFOrdEqual + OpSignBitSet

Signed-off-by: Sidorov, Dmitry <dmitry.sidorov@intel.com>

MrSidims commented on lib/SPIRV/SPIRVWriter.cpp:
```diff
@@ -4245,6 +4245,96 @@ SPIRVValue *LLVMToSPIRVBase::transIntrinsicInst(IntrinsicInst *II,
+    }
+    if (FPClass & fcNegSubnormal || FPClass & fcPosSubnormal) {
+      // If value is both not NaN and Normal, then it's subnormal
+      // TODO: it's quite silly check, can we figure out something better?
```
@svenvh @asudarsa @bashbaug may I ask you to take a look? I assumed, that subnormal test should be test for not Nan and not Normal, though may be you see better ways to perform that.

jcranmer-intel replied:
Simplest check here is to check if the exponent bits are minimum exponent, and mantissa is nonzero. Looking at the TargetLowering variant, this can be reduced further to `unsigned(mask_sign(V) - 1) < all_ones_mantissa`

MrSidims commented on lib/SPIRV/SPIRVWriter.cpp:
```diff
@@ -4245,6 +4245,96 @@ SPIRVValue *LLVMToSPIRVBase::transIntrinsicInst(IntrinsicInst *II,
+    if (FPClass & fcNegZero || FPClass & fcPosZero) {
+      // Map on OpFOrdEqual compare to 0.0 with following check for sign bit
+      Constant *Zero = ConstantFP::getZero(II->getArgOperand(0)->getType());
+      auto *TestIsZero = BM->addCmpInst(OpFOrdEqual, Ty, InputFloat,
```
Adding OpFOrdEqual is also a bit questionable, since it's instruction from SPIR-V 1.6 and we officially support only SPIR-V 1.4 in the translator.

Commit:
Address review

Signed-off-by: Sidorov, Dmitry <dmitry.sidorov@intel.com>

jcranmer-intel commented on lib/SPIRV/SPIRVWriter.cpp:
```diff
@@ -4245,6 +4245,96 @@ SPIRVValue *LLVMToSPIRVBase::transIntrinsicInst(IntrinsicInst *II,
+    // Since a single intrinsic can provide multiple tests - here we might end
+    // up adding several sequences of SPIR-V instructions
+    if (FPClass & fcSNan || FPClass & fcQNan) {
+      // There are no signaling and quite NaN instructions in SPIR-V, so map
+      // both of the tests on OpIsNan
+      ResultVec.emplace_back(BM->addInstTemplate(OpIsNan, Ops, BB, Ty));
```
It's not correct to consider the tests for sNaN or qNaN as equivalent, because the definition is based on the bit pattern (and InstCombine may canonicalize such a test to an fpclass check!). Distinguishing between qNaN and sNaN requires checking if the quiet bit (MSB of the mantissa) is set or not.

jcranmer-intel reviewed (commented):
Since I've been reviewing several of the upstream patches related here, I felt I should give some drive-by comments.

The "correct" lowering of is_fpclass is given by TargetLowering::expandIS_FPCLASS  (https://github.com/llvm/llvm-project/blob/main/llvm/lib/CodeGen/SelectionDAG/TargetLowering.cpp#L8019). Instcombine should be canonicalizing the tests that are equivalent to a single comparison into a single instruction, so checking for the really simple cases may not be all that beneficial.

The general strategy for the implementation should not to be to check with floating-point operations, but integer operations on bitcast-to-int. This avoids issues where floating-point operations may have underspecified issues (e.g., comparison to zero returning different results depending on denormals-are-zero setting).

MrSidims commented:
@jcranmer-intel thanks! I've applied the comments. Not QNan, SNan, PosSubnormal, NegSubnormal, PosZero and NegZero are checked after bitcast to int. Checks for normal and inf remained to be the same - relying on existing SPIR-V instructions.

MrSidims commented on lib/SPIRV/SPIRVWriter.cpp:
```diff
@@ -4245,6 +4245,151 @@ SPIRVValue *LLVMToSPIRVBase::transIntrinsicInst(IntrinsicInst *II,
+    // Since a single intrinsic can provide multiple tests - here we might end
+    // up adding several sequences of SPIR-V instructions
+    if (FPClass & fcNan) {
+      auto *TestIsNan = BM->addInstTemplate(OpIsNan, {InputFloat->getId()}, BB,
```
To move it a bit later in the code

Commit:
Fix bug and add test

Signed-off-by: Sidorov, Dmitry <dmitry.sidorov@intel.com>

MrSidims commented on lib/SPIRV/SPIRVWriter.cpp:
```diff
@@ -4245,6 +4245,151 @@ SPIRVValue *LLVMToSPIRVBase::transIntrinsicInst(IntrinsicInst *II,
+        ResultVec.emplace_back(TestIsInf);
+      else
+        // Map on OpIsInf with following check for sign bit
+        ResultVec.emplace_back(GetNegPosInstTest(TestIsInf,
+                                                 FPClass & fcNegInf));
```
Here and after logical error: check for bit should be done on float, not on IsTest result

Commit:
clang-format

Signed-off-by: Sidorov, Dmitry <dmitry.sidorov@intel.com>

MrSidims commented:
@jcranmer-intel @asudarsa @svenvh @maksimsab @jgstarIntel please take a look

MrSidims commented on lib/SPIRV/SPIRVWriter.cpp:
```diff
@@ -4245,6 +4245,153 @@ SPIRVValue *LLVMToSPIRVBase::transIntrinsicInst(IntrinsicInst *II,
+    std::vector<SPIRVValue *> ResultVec;
+
+    // Adds test for Negative/Positive values
+    auto GetNegPosInstTest = [&](SPIRVValue *TestInst, bool IsNegative) {
```
Hm, I feel like there is room for optimization(s): for example test for sign can be cached at first entry and its ID can be reused by other tests.

asudarsa commented on test/llvm-intrinsics/fpclass.ll:
```diff
@@ -0,0 +1,451 @@
+; RUN: spirv-val %t.spv
+; RUN: llvm-spirv -r %t.spv -o %t.rev.bc
+
+; Just check that reverse translation works
```
Can we have a quick check to ensure that fpclass is not generated here?

Thanks

MrSidims replied:
I guess it's being checked implicitly by the fact, that forward translation doesn't error out due to unknown intrinsic

asudarsa commented on lib/SPIRV/SPIRVWriter.cpp:
```diff
@@ -4245,6 +4245,153 @@ SPIRVValue *LLVMToSPIRVBase::transIntrinsicInst(IntrinsicInst *II,
+        ResultVec.emplace_back(
+            GetNegPosInstTest(TestIsInf, FPClass & fcNegInf));
+    }
+    if (FPClass & fcNormal) {
```
Is there a possibility that fcNormal bit is set and both fcPosNormal and fcNegNormal are not set? In such such, what is the expected behavior? Thanks

asudarsa reviewed (approved):
Overall LGTM. Just a couple of minor questions/comments.

Thanks

Commit:
Add inverted test

Signed-off-by: Sidorov, Dmitry <dmitry.sidorov@intel.com>

MrSidims commented:
@asudarsa may I ask to take another look? I've added 2 enhancements: added caching for sign test results to be reused => less IsSignBit instructions generated and added check for inverted test aka (is inf, is normal, is subnormal is zero) = (is not nan)

Commit:
clang-format and clang-tidy

Signed-off-by: Sidorov, Dmitry <dmitry.sidorov@intel.com>

MrSidims commented:
@jcranmer-intel may I ask you to revisit the PR?

jcranmer-intel reviewed (approved):
In theory, the tests could be a little more comprehensive, but without an `update-test-checks`-like script, it is going to be too painful to make such a test for the limited extra value it would bring.

MrSidims commented on lib/SPIRV/SPIRVWriter.cpp:
```diff
@@ -4318,6 +4245,204 @@ SPIRVValue *LLVMToSPIRVBase::transIntrinsicInst(IntrinsicInst *II,
+      IsInverted = true;
+      FPClass = InvertedCheck;
+    }
+    auto GetInvertedTestIfNeeded = [&](SPIRVValue *TestInst) -> SPIRVValue * {
```
Thanks! I'll open an issue and investigate this concern a bit later.

MrSidims replied:
(note, there were some follow up PRs, may be they resolve the issue, may be not, requires me to double check it :) )

Created https://github.com/KhronosGroup/SPIRV-LLVM-Translator/issues/3696
